File: docs/calculator.js
// Estado da aplicação
let classes = [];
let contadorClasses = 0;

// Elementos DOM
const aporteTotalInput = document.getElementById('aporte-total');
const classesContainer = document.getElementById('classes-container');
const addClasseBtn = document.getElementById('add-classe');
const exemploPresetBtn = document.getElementById('exemplo-preset');
const calcularBtn = document.getElementById('calcular');
const limparBtn = document.getElementById('limpar');
const resultadoDiv = document.getElementById('resultado');
const totalPercentualSpan = document.getElementById('total-percentual');
const warningPercentual = document.getElementById('warning-percentual');

// Inicialização
document.addEventListener('DOMContentLoaded', function () {
    // Adicionar eventos
    addClasseBtn.addEventListener('click', function () {
        adicionarClasse();
    });
    exemploPresetBtn.addEventListener('click', carregarExemplo);
    calcularBtn.addEventListener('click', calcularAporte);
    limparBtn.addEventListener('click', limparTudo);

    // Inicializar sem exemplo - página limpa
    atualizarTotalPercentual();
});

function adicionarClasse(nome = '', valorAtual = '', metaPercentual = '') {
    contadorClasses++;
    const classeId = `classe-${contadorClasses}`;

    const classeDiv = document.createElement('div');
    classeDiv.className = 'classe-item';
    classeDiv.id = classeId;

    classeDiv.innerHTML = `
        <div class="classe-header">
            <div class="classe-numero">${contadorClasses}</div>
            <button type="button" class="remove-classe" onclick="removerClasse('${classeId}')">×</button>
        </div>
        <div class="classe-fields">
            <div class="form-group">
                <label>Nome da Classe:</label>
                <input type="text" class="nome-classe" placeholder="Ex: Renda Fixa, Ações Brasil..." value="${nome}">
            </div>
            <div class="form-group">
                <label>Valor Atual (R$):</label>
                <input type="number" class="valor-atual" min="0" step="0.01" placeholder="0.00" value="${valorAtual}">
            </div>
            <div class="form-group">
                <label>Meta (%):</label>
                <input type="number" class="meta-percentual" min="0" max="100" step="0.01" placeholder="0.0" value="${metaPercentual}">
            </div>
        </div>
    `;

    classesContainer.appendChild(classeDiv);

    // Adicionar eventos de mudança para atualizar total
    const inputs = classeDiv.querySelectorAll('input');
    inputs.forEach(input => {
        input.addEventListener('input', atualizarTotalPercentual);
    });

    atualizarTotalPercentual();
}

function removerClasse(classeId) {
    const classeDiv = document.getElementById(classeId);
    if (classeDiv) {
        classeDiv.remove();
        renumerarClasses();
        atualizarTotalPercentual();
    }
}

function renumerarClasses() {
    const classeItems = document.querySelectorAll('.classe-item');
    contadorClasses = classeItems.length;

    classeItems.forEach((item, index) => {
        const numeroDiv = item.querySelector('.classe-numero');
        if (numeroDiv) {
            numeroDiv.textContent = index + 1;
        }

        // Atualizar o ID e onclick do botão remover
        const novoId = `classe-${index + 1}`;
        item.id = novoId;
        const removeBtn = item.querySelector('.remove-classe');
        if (removeBtn) {
            removeBtn.setAttribute('onclick', `removerClasse('${novoId}')`);
        }
    });
}

function atualizarTotalPercentual() {
    const metasInputs = document.querySelectorAll('.meta-percentual');
    let total = 0;

    metasInputs.forEach(input => {
        const valor = parseFloat(input.value) || 0;
        total += valor;
    });

    totalPercentualSpan.textContent = total.toFixed(1);

    if (Math.abs(total - 100) > 0.01) {
        warningPercentual.classList.remove('hidden');
        totalPercentualSpan.style.color = '#dc3545';
    } else {
        warningPercentual.classList.add('hidden');
        totalPercentualSpan.style.color = '#28a745';
    }
}

function carregarExemplo() {
    // Limpar classes existentes
    classesContainer.innerHTML = '';
    contadorClasses = 0;

    // Adicionar exemplo padrão
    adicionarClasse('CDBs', '2240.00', '65.0');
    adicionarClasse('ETFs', '1100.00', '35.0');

    // Definir aporte exemplo
    aporteTotalInput.value = '800.00';
}

function limparTudo() {
    classesContainer.innerHTML = '';
    contadorClasses = 0;
    aporteTotalInput.value = '';
    resultadoDiv.classList.add('hidden');
    atualizarTotalPercentual();
}

function coletarDados() {
    const aporteMensalTotal = parseFloat(aporteTotalInput.value) || 0;
    const classesAtivos = [];

    const classeItems = document.querySelectorAll('.classe-item');

    classeItems.forEach(item => {
        const nome = item.querySelector('.nome-classe').value.trim();
        const valorAtual = parseFloat(item.querySelector('.valor-atual').value) || 0;
        const metaPercentual = parseFloat(item.querySelector('.meta-percentual').value) || 0;

        if (nome) {
            classesAtivos.push({
                nome: nome,
                valorAtual: valorAtual,
                metaPercentual: metaPercentual
            });
        }
    });

    return {
        aporteMensalTotal: aporteMensalTotal,
        classesAtivos: classesAtivos
    };
}

function validarDados(dados) {
    const erros = [];

    if (dados.aporteMensalTotal <= 0) {
        erros.push('Aporte mensal deve ser maior que zero');
    }

    if (dados.classesAtivos.length === 0) {
        erros.push('Adicione pelo menos uma classe de ativo');
    }

    const totalPercentual = dados.classesAtivos.reduce((total, classe) => total + classe.metaPercentual, 0);
    if (Math.abs(totalPercentual - 100) > 0.01) {
        erros.push(`Os percentuais das metas somam ${totalPercentual.toFixed(1)}% ao invés de 100%`);
    }

    return erros;
}

function calcularAporteAlgoritmo(classesAtivos, aporteTotal) {
    // Implementação do mesmo algoritmo do Python
    const totalAtualCarteira = classesAtivos.reduce((total, classe) => total + classe.valorAtual, 0);
    const totalFinalCarteira = totalAtualCarteira + aporteTotal;

    const aportes = [];
    let aporteRestante = aporteTotal;

    // Calcular necessidades para cada classe
    const necessidades = classesAtivos.map(classe => {
        const valorIdeal = totalFinalCarteira * (classe.metaPercentual / 100);
        const necessidade = valorIdeal - classe.valorAtual;

        return {
            nome: classe.nome,
            valorAtual: classe.valorAtual,
            metaPercentual: classe.metaPercentual,
            valorIdeal: valorIdeal,
            necessidade: necessidade
        };
    });

    // Primeira passada: distribuir aporte baseado nas necessidades
    necessidades.forEach(necessidade => {
        let aporteClasse = 0;

        if (aporteRestante > 0 && necessidade.necessidade > 0) {
            aporteClasse = Math.min(aporteRestante, necessidade.necessidade);
            aporteRestante -= aporteClasse;
        }

        aportes.push({
            nome: necessidade.nome,
            valorAtual: necessidade.valorAtual,
            metaPercentual: necessidade.metaPercentual,
            aporte: aporteClasse,
            novoValor: necessidade.valorAtual + aporteClasse
        });
    });

    // Se ainda sobrou aporte, distribuir proporcionalmente
    if (aporteRestante > 0) {
        const totalPercentual = classesAtivos.reduce((total, classe) => total + classe.metaPercentual, 0);

        aportes.forEach(aporte => {
            const proporcao = aporte.metaPercentual / totalPercentual;
            const aporteAdicional = aporteRestante * proporcao;
            aporte.aporte += aporteAdicional;
            aporte.novoValor += aporteAdicional;
        });
    }

    return {
        totalAtualCarteira: totalAtualCarteira,
        totalFinalCarteira: totalFinalCarteira,
        aportes: aportes
    };
}

function calcularAporte() {
    const dados = coletarDados();
    const erros = validarDados(dados);

    if (erros.length > 0) {
        alert('Erros encontrados:\n' + erros.join('\n'));
        return;
    }

    const resultado = calcularAporteAlgoritmo(dados.classesAtivos, dados.aporteMensalTotal);
    exibirResultado(resultado);
}

function exibirResultado(resultado) {
    // Carteira ANTES
    const carteiraAntes = document.getElementById('carteira-antes');
    carteiraAntes.innerHTML = '';

    resultado.aportes.forEach(aporte => {
        const percentualAtual = resultado.totalAtualCarteira > 0 ?
            (aporte.valorAtual / resultado.totalAtualCarteira * 100) : 0;

        const div = document.createElement('div');
        div.className = 'classe-resultado';
        div.innerHTML = `
            <span><strong>${aporte.nome}:</strong> R$ ${aporte.valorAtual.toFixed(2)}</span>
            <span>(${percentualAtual.toFixed(1)}%)</span>
        `;
        carteiraAntes.appendChild(div);
    });

    document.getElementById('total-antes').textContent = resultado.totalAtualCarteira.toFixed(2);

    // Recomendação de APORTE
    const recomendacaoAporte = document.getElementById('recomendacao-aporte');
    recomendacaoAporte.innerHTML = '';

    resultado.aportes.forEach(aporte => {
        const div = document.createElement('div');
        div.className = aporte.aporte > 0 ? 'aporte-item' : 'aporte-item zero';

        const texto = aporte.aporte > 0 ?
            `${aporte.nome}: R$ ${aporte.aporte.toFixed(2)}` :
            `${aporte.nome}: R$ 0.00 (não necessário)`;

        div.innerHTML = `<span>${texto}</span>`;
        recomendacaoAporte.appendChild(div);
    });

    const totalAporte = resultado.aportes.reduce((total, aporte) => total + aporte.aporte, 0);
    document.getElementById('total-aporte-calculado').textContent = totalAporte.toFixed(2);

    // Carteira DEPOIS
    const carteiraDepois = document.getElementById('carteira-depois');
    carteiraDepois.innerHTML = '';

    resultado.aportes.forEach(aporte => {
        const percentualNovo = (aporte.novoValor / resultado.totalFinalCarteira * 100);
        const meta = aporte.metaPercentual;
        const status = Math.abs(percentualNovo - meta) < 1 ? 'OK' : 'DESBALANCEADO';
        const classeStatus = Math.abs(percentualNovo - meta) < 1 ? '' : 'warning';

        const div = document.createElement('div');
        div.className = `classe-resultado ${classeStatus}`;
        div.innerHTML = `
            <span><strong>${aporte.nome}:</strong> R$ ${aporte.novoValor.toFixed(2)}</span>
            <span>(${percentualNovo.toFixed(1)}% | Meta: ${meta.toFixed(1)}%) [${status}]</span>
        `;
        carteiraDepois.appendChild(div);
    });

    document.getElementById('total-depois').textContent = resultado.totalFinalCarteira.toFixed(2);

    // Mostrar resultado
    resultadoDiv.classList.remove('hidden');
    resultadoDiv.scrollIntoView({ behavior: 'smooth' });
}

File: cli/calculator.py
import json
import os
from datetime import datetime

NOME_ARQUIVO_DADOS = "dados_carteira.json"


def carregar_dados_json():
    """Lê o arquivo de dados .json e retorna como um dicionário Python."""
    if not os.path.exists(NOME_ARQUIVO_DADOS):
        raise FileNotFoundError(f"Arquivo '{NOME_ARQUIVO_DADOS}' não encontrado. Crie o arquivo com os dados iniciais.")

    with open(NOME_ARQUIVO_DADOS, 'r', encoding='utf-8') as f:
        dados = json.load(f)
        
    total_percentual = sum(classe["metaPercentual"] for classe in dados["classesAtivos"])
    if abs(total_percentual - 100.0) > 0.01:
        print(f"AVISO: Os percentuais das metas somam {total_percentual}% ao invés de 100%!")
        
    return dados


def salvar_dados_json(dados_carteira):
    """Gera e salva o novo conteúdo no arquivo de dados .json para o próximo mês."""
    dados_para_salvar = {
        "_comment": f"Valores atualizados automaticamente em {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}. No proximo mes, atualize estes saldos antes de rodar o script.",
        "aporteMensalTotal": dados_carteira["aporteMensalTotal"],
        "classesAtivos": []
    }
    
    # Copiar classes de ativos com valores atualizados
    for classe in dados_carteira["classesAtivos"]:
        dados_para_salvar["classesAtivos"].append({
            "nome": classe["nome"],
            "valorAtual": round(classe["valorAtual"], 2),
            "metaPercentual": classe["metaPercentual"]
        })
    
    with open(NOME_ARQUIVO_DADOS, 'w', encoding='utf-8') as f:
        json.dump(dados_para_salvar, f, indent=4, ensure_ascii=False)
    print(f"\nAVISO: O arquivo '{NOME_ARQUIVO_DADOS}' foi atualizado com os novos totais!")


def calcular_aporte(classes_ativos, aporte_total):
    """Calcula quanto aportar em cada classe de ativo para rebalancear a carteira."""
    
    # Calcular totais atuais
    total_atual_carteira = sum(classe["valorAtual"] for classe in classes_ativos)
    total_final_carteira = total_atual_carteira + aporte_total
    
    # Calcular aportes necessários para cada classe
    aportes = []
    aporte_restante = aporte_total
    
    # Calcular o quanto cada classe precisa para atingir a meta
    necessidades = []
    for classe in classes_ativos:
        valor_ideal = total_final_carteira * (classe["metaPercentual"] / 100)
        necessidade = valor_ideal - classe["valorAtual"]
        necessidades.append({
            "nome": classe["nome"],
            "valorAtual": classe["valorAtual"],
            "metaPercentual": classe["metaPercentual"],
            "valorIdeal": valor_ideal,
            "necessidade": necessidade
        })
    
    # Calcular aportes respeitando o limite do aporte total
    for i, necessidade in enumerate(necessidades):
        if aporte_restante <= 0:
            aporte_classe = 0
        elif necessidade["necessidade"] <= 0:
            aporte_classe = 0
        else:
            aporte_classe = min(aporte_restante, necessidade["necessidade"])
        
        aporte_restante -= aporte_classe
        
        aportes.append({
            "nome": necessidade["nome"],
            "valorAtual": necessidade["valorAtual"],
            "metaPercentual": necessidade["metaPercentual"],
            "aporte": aporte_classe,
            "novoValor": necessidade["valorAtual"] + aporte_classe
        })
    
    # Se ainda sobrou aporte, distribuir proporcionalmente
    if aporte_restante > 0:
        total_percentual = sum(classe["metaPercentual"] for classe in classes_ativos)
        for aporte in aportes:
            proporcao = aporte["metaPercentual"] / total_percentual
            aporte_adicional = aporte_restante * proporcao
            aporte["aporte"] += aporte_adicional
            aporte["novoValor"] += aporte_adicional
    
    return {
        "totalAtualCarteira": total_atual_carteira,
        "totalFinalCarteira": total_final_carteira,
        "aportes": aportes
    }


if __name__ == "__main__":
    dados_carteira = carregar_dados_json()

    resultado = calcular_aporte(
        dados_carteira["classesAtivos"],
        dados_carteira["aporteMensalTotal"]
    )

    print("=" * 50)
    print(" ANÁLISE DE APORTE MENSAL")
    print("=" * 50)
    
    # Mostrar carteira ANTES do aporte
    print("\nCARTEIRA ANTES DO APORTE:\n")
    for aporte in resultado["aportes"]:
        percentual_atual = (aporte["valorAtual"] / resultado["totalAtualCarteira"] * 100) if resultado["totalAtualCarteira"] > 0 else 0
        print(f"   {aporte['nome']}: R$ {aporte['valorAtual']:.2f} ({percentual_atual:.1f}%)")
    print(f"\nPATRIMÔNIO TOTAL: R$ {resultado['totalAtualCarteira']:.2f}\n")
    
    # Mostrar recomendação de aportes
    print("-" * 50)
    print("\n🎯 RECOMENDAÇÃO DE APORTE:\n")
    total_aporte = sum(aporte["aporte"] for aporte in resultado["aportes"])
    for aporte in resultado["aportes"]:
        if aporte["aporte"] > 0:
            print(f"   ➡️  {aporte['nome']}: R$ {aporte['aporte']:.2f}")
        else:
            print(f"   ⚪  {aporte['nome']}: R$ 0.00 (não necessário)")
    print(f"\nTOTAL APORTADO: R$ {total_aporte:.2f}\n")

    
    # Mostrar carteira DEPOIS do aporte
    print("-" * 50)
    print("\nCARTEIRA DEPOIS DO APORTE:\n")
    for aporte in resultado["aportes"]:
        percentual_novo = (aporte["novoValor"] / resultado["totalFinalCarteira"] * 100)
        meta = aporte["metaPercentual"]
        status = "✅" if abs(percentual_novo - meta) < 1 else "⚠️"
        print(f"   {aporte['nome']}: R$ {aporte['novoValor']:.2f} ({percentual_novo:.1f}% | Meta: {meta:.1f}%) {status}")
    print(f"\nNOVO PATRIMÔNIO TOTAL: R$ {resultado['totalFinalCarteira']:.2f}")
    
    print("=" * 50)

    # Atualizar dados para salvar
    for i, aporte in enumerate(resultado["aportes"]):
        dados_carteira["classesAtivos"][i]["valorAtual"] = aporte["novoValor"]
    
    salvar_dados_json(dados_carteira)